### PR TITLE
Fix all DScanner warnings + use -return-exit-code re-enable it

### PR DESCRIPTION
**Pull request by wilzbach**
tl;dr: DScanner started to segfaulted in December due to the introduction of the AscendingPageAllocator, which used `~this(){}`. The DScanner bug has been fixed and now we can fix the DScanner warnings that sneaked in and re-enable in.

See also: https://github.com/dlang/phobos/pull/6148

### CODE REVIEW

File: std/net/curl.d [hunk 1]
```diff
@@ -313,7 +313,7 @@ version(StdUnittest)
 version(StdDdoc) import std.stdio;
 
 // Default data timeout for Protocols
-enum _defaultDataTimeout = dur!"minutes"(2);
+private enum _defaultDataTimeout = dur!"minutes"(2);
```
Comment (wilzbach) [inline std/net/curl.d]: Undocumented public symbol, it's not too late for this one. It has been only added recently: https://github.com/dlang/phobos/commit/91eaa508c4

File: std/traits.d [hunk 2]
```diff
@@ -5238,7 +5238,8 @@ template IntegralTypeOf(T)
             static assert( is(Q!T == IntegralTypeOf!( SubTypeOf!(Q!T) )));
         }
 
-    static foreach (T; AliasSeq!(void, bool, FloatingPointTypeList, /*ImaginaryTypeList, ComplexTypeList,*/ CharTypeList))
+    static foreach (T; AliasSeq!(void, bool, FloatingPointTypeList,
+                /*ImaginaryTypeList, ComplexTypeList,*/ CharTypeList))
```
Comment (wilzbach) [inline std/traits.d]: Long line

File: posix.mak [hunk 3]
```diff
@@ -535,7 +535,7 @@ style: publictests style_lint
 # runs static code analysis with Dscanner
 dscanner: | $(DSCANNER_DIR)/dsc
 	@echo "Running DScanner"
-	$(DEBUGGER) -q -ex run -ex bt -batch --args $(DSCANNER_DIR)/dsc --config .dscanner.ini --styleCheck etc std -I.
+	$(DEBUGGER) -return-child-result -q -ex run -ex bt -batch --args $(DSCANNER_DIR)/dsc --config .dscanner.ini --styleCheck etc std -I.
```
Comment (wilzbach) [inline posix.mak]: FYI: DScanner doesn't seem to be super stable, so imho it's better to keep running it within GDB for a while. The overhead isn't too big and we have done so for the last two months.

**Comment by wilzbach** [comment]
image

Yeah, this is finally ready. Sorry for the commit noise.

File: std/algorithm/iteration.d [hunk 4]
```diff
@@ -4689,6 +4689,47 @@ if (isSomeChar!C)
...
+    static assert(hasDifferentAutodecoding!(dstring, AliasSeq!(wstring, wstring)));
+
+    // both (range, needle) need auto-decoding and don't share the same common type
+    static foreach (T; AliasSeq!(string, wstring, dstring))
+    {
+        static assert(hasDifferentAutodecoding!(T, AliasSeq!(wstring, string)));
+        static assert(hasDifferentAutodecoding!(T, AliasSeq!(dstring, string)));
+        static assert(hasDifferentAutodecoding!(T, AliasSeq!(wstring, dstring)));
+    }
+}
```
[the remark this answers is not in the document]
Comment (JackStouffer) [reply std/algorithm/iteration.d]: I don't understand. This symbol is not documented and the unit test is not an example.

[the remark this answers is not in the document]
Comment (wilzbach) [reply std/algorithm/iteration.d]: The warning was for `template substitute`. I don't remember which one, but one of the documentation engines used to dislike different symbols (even if they are private) between declaration and unittest. It looks like this has been fixed, so I can look into updating DScanner... though maybe we can live with this workaround until DScanner has been upgraded?

**Pull request merged**